nssp/delphi_nssp/patch.py

```diff
@@ -0,0 +1,195 @@
+        except IOError:
+            logger.warning(
+                "Source backup for this date does not exist on the remote server.", missing_filename=remote_file_name
+            )
+    sftp.close()
+    ssh.close()
+
+    if num_files_transferred == 0:
+        logger.error("No source data was transferred. Check the source backup server for potential issues.")
+        sys.exit(1)
+
+
+def good_patch_config(params, logger):
+    """
+    Check if the params.json file is correctly configured for patching.
+
+    params: Dict[str, Any]
+        Nested dictionary of parameters, typically loaded from params.json file.
+    logger: Logger object
+        Logger object to log messages.
+    """
+    valid_config = True
+    custom_run = params["common"].get("custom_run", False)
+    if not custom_run:
+        logger.error("Calling patch.py without custom_run flag set true.")
+        valid_config = False
+
+    patch_config = params.get("patch", {})
+    if patch_config == {}:
+        logger.error("Custom flag is on, but patch section is missing.")
+        valid_config = False
+    else:
+        required_patch_keys = ["start_issue", "end_issue", "patch_dir", "source_dir"]
+        missing_keys = [key for key in required_patch_keys if key not in patch_config]
+        if missing_keys:
+            logger.error("Patch section is missing required key(s)", missing_keys=missing_keys)
+            valid_config = False
+        else:
+            try:  # issue dates validity check
+                start_issue = datetime.strptime(patch_config["start_issue"], "%Y-%m-%d")
+                end_issue = datetime.strptime(patch_config["end_issue"], "%Y-%m-%d")
+                if start_issue > end_issue:
+                    logger.error("Start issue date is after end issue date.")
+                    valid_config = False
+            except ValueError:
+                logger.error("Issue dates must be in YYYY-MM-DD format.")
+                valid_config = False
+
+            if not path.isdir(patch_config["source_dir"]):
+                logger.error("Source directory does not exist.", source_dir=patch_config["source_dir"])
+                valid_config = False
+
+    if valid_config:
+        logger.info("Good patch configuration.")
+        return True
+    logger.info("Bad patch configuration.")
+    return False
+
+
+def patch():
+    """
+    Run the nssp indicator for a range of issue dates.
+
+    The range of issue dates is specified in params.json using the following keys:
+    - "patch": Only used for patching data
+        - "source_backup_credentials": (Optional)
+        Add this object to download source data from a backup server to source_dir.
+        Remove to assume source csv are already in source_dir.
+            - "host": str, hostname of the server where source data is backed up
+            - "user": str, username to log in to the server
+            - "path": str, path to the directory containing backup csv files
```

Comment:
I keep this optional because downloading stashed source files take _awhile_, so if for some reason you're re-runing patch for the same period again and again, removing this config chunk would save time.

Reply:
Probably a good idea to check if the files/directory is not empty at the beginning of the get_source_data. Then even if it was called, it won't make the calls over and over agin

Reply:
fixed at [7e06f94](https://github.com/cmu-delphi/covidcast-indicators/pull/2000/commits/7e06f94d7b6a9668ad8e9843e6838f5c8c369a63).